File: node/Assembler/src/templateEngine/enginePreProcess.js
/**
 * PreProcess template engine implementation that only does merging using preprocessed data structures
 * All parsing is done by TemplateLoader, this engine only handles merging
 */
export class EnginePreProcess {
    constructor(appViewPrefix = '') {
        this.appViewPrefix = appViewPrefix;
    }

    /**
     * Merges templates using preprocessed data structures
     * This method only does merging using preprocessed data structures - no loading or parsing
     * @param {string} appSite The application site name for template key generation
     * @param {string} appFile The application file name
     * @param {string|null} appView The application view name (optional)
     * @param {Object.<string, PreprocessedTemplate>} preprocessedTemplates Dictionary of preprocessed templates for this specific appSite
     * @param {boolean} enableJsonProcessing Whether to enable JSON data processing
     * @returns {string} HTML with placeholders replaced using preprocessed structures
     */
    mergeTemplates(appSite, appFile, appView, preprocessedTemplates, enableJsonProcessing = true) {
        if (!preprocessedTemplates || 
            (preprocessedTemplates instanceof Map && preprocessedTemplates.size === 0) ||
            (!(preprocessedTemplates instanceof Map) && Object.keys(preprocessedTemplates).length === 0)) {
            return '';
        }

        // Use the getTemplate method to retrieve the main template
        const mainPreprocessed = this.getTemplate(appSite, appFile, preprocessedTemplates, appView, this.appViewPrefix, true);
        if (!mainPreprocessed) {
            return '';
        }

        // Start with original content
        let contentHtml = mainPreprocessed.originalContent;

        // Apply ALL replacement mappings from ALL templates (TemplateLoader did all the processing)
        contentHtml = this.applyTemplateReplacements(contentHtml, preprocessedTemplates, enableJsonProcessing, appView);

        return contentHtml;
    }

    /**
     * Retrieves a template from the preprocessed templates dictionary based on various scenarios including AppView fallback logic
     * @param {string} appSite The application site name
     * @param {string} templateName The template name (can be appFile or placeholderName)
     * @param {Object.<string, PreprocessedTemplate>} preprocessedTemplates Dictionary of preprocessed templates
     * @param {string|null} appView The application view name (optional)
     * @param {string|null} appViewPrefix The application view prefix (optional, uses instance property if not provided)
     * @param {boolean} useAppViewFallback Whether to apply AppView fallback logic
     * @returns {PreprocessedTemplate|null} The template if found, null otherwise
     */
    getTemplate(appSite, templateName, preprocessedTemplates, appView = null, appViewPrefix = null, useAppViewFallback = true) {
        if (!preprocessedTemplates || 
            (preprocessedTemplates instanceof Map && preprocessedTemplates.size === 0) ||
            (!(preprocessedTemplates instanceof Map) && Object.keys(preprocessedTemplates).length === 0)) {
            return null;
        }

        const viewPrefix = appViewPrefix || this.appViewPrefix;
        
        // FIRST: Check for AppView-specific template resolution when AppView context is provided
        if (useAppViewFallback && appView && viewPrefix && templateName.toLowerCase().includes(viewPrefix.toLowerCase())) {
            // Direct replacement: Replace the AppViewPrefix with the AppView value
            // For example: Html3AContent with AppViewPrefix=Html3A and AppView=html3B becomes html3BContent
            const appKey = this.replaceAllCaseInsensitive(templateName, viewPrefix, appView);
            const fallbackTemplateKey = `${appSite.toLowerCase()}_${appKey.toLowerCase()}`;
            
            if (preprocessedTemplates instanceof Map) {
                if (preprocessedTemplates.has(fallbackTemplateKey)) {
                    return preprocessedTemplates.get(fallbackTemplateKey);
                }
            } else if (preprocessedTemplates[fallbackTemplateKey]) {
                return preprocessedTemplates[fallbackTemplateKey]; // Found AppView-specific template, use it
            }
        }

        // SECOND: If no AppView-specific template found, try primary template
        const primaryTemplateKey = `${appSite.toLowerCase()}_${templateName.toLowerCase()}`;
        if (preprocessedTemplates instanceof Map) {
            if (preprocessedTemplates.has(primaryTemplateKey)) {
                return preprocessedTemplates.get(primaryTemplateKey);
            }
        } else if (preprocessedTemplates[primaryTemplateKey]) {
            return preprocessedTemplates[primaryTemplateKey];
        }
        return null;
    }

    /**
     * Applies all replacement mappings from all templates - NO processing logic, only direct replacements
     * @param {string} content The content to apply replacements to
     * @param {Object.<string, PreprocessedTemplate>} preprocessedTemplates Dictionary of preprocessed templates
     * @param {boolean} enableJsonProcessing Whether to enable JSON processing
     * @param {string|null} appView The application view name
     * @returns {string} Content with all replacements applied
     */
    applyTemplateReplacements(content, preprocessedTemplates, enableJsonProcessing, appView) {
        let result = content;

        // Apply replacement mappings from all templates in multiple passes until no more changes
        const maxPasses = 10; // Prevent infinite loops
        let currentPass = 0;
        
        do {
            const previous = result;
            currentPass++;
            
            // Apply replacement mappings from all templates
            const templates = preprocessedTemplates instanceof Map ? 
                Array.from(preprocessedTemplates.values()) : 
                Object.values(preprocessedTemplates);
                
            for (const template of templates) {
                // CRITICAL: Apply slotted template mappings FIRST (before JSON processing changes the content)
                for (const mapping of template.replacementMappings) {
                    if ((mapping.type === 2 || mapping.type === "SlottedTemplate") && result.includes(mapping.originalText)) {
                        result = result.replaceAll(mapping.originalText, mapping.replacementText);
                    }
                }
                
                // Then apply other replacement mappings (simple templates) with AppView logic
                for (const mapping of template.replacementMappings) {
                    if ((mapping.type === 1 || mapping.type === "SimpleTemplate") && result.includes(mapping.originalText)) {
                        // Apply AppView logic before replacement
                        const replacementText = this.applyAppViewLogicToReplacement(mapping.originalText, mapping.replacementText, preprocessedTemplates, appView);
                        result = result.replaceAll(mapping.originalText, replacementText);
                    }
                }
                
                // Apply JSON replacement mappings only if JSON processing is enabled
                if (enableJsonProcessing) {
                    for (const mapping of template.replacementMappings) {
                        if ((mapping.type === 0 || mapping.type === "JsonPlaceholder") && result.includes(mapping.originalText)) {
                            result = result.replaceAll(mapping.originalText, mapping.replacementText);
                        }
                    }
                }
                
                // Apply JSON placeholders if JSON processing is enabled (LAST)
                if (enableJsonProcessing) {
                    for (const placeholder of template.jsonPlaceholders) {
                        result = this.replaceAllCaseInsensitive(result, placeholder.placeholder, placeholder.value);
                    }
                }
            }
            
            // Break if no changes were made or max passes reached
            if (result === previous || currentPass >= maxPasses) {
                break;
            }
        } while (true);

        return result;
    }

    /**
     * Handle conditional blocks like {{@Selected}}...{{/Selected}}
     * Ported from C# HandleConditional method
     * @param {string} input Input string
     * @param {string} key Conditional key (e.g., "Selected")
     * @param {boolean} condition Whether to include or remove the conditional content
     * @returns {string} String with conditional blocks processed
     */
    handleConditional(input, key, condition) {
        // Support spaces inside block tags, e.g. {{@Selected}} ... {{ /Selected}}
        const condStart = `{{@${key}}}`;
        const condEnd = `{{ /${key}}}`;
        
        let startIdx = input.toLowerCase().indexOf(condStart.toLowerCase());
        let endIdx = input.toLowerCase().indexOf(condEnd.toLowerCase());
        
        while (startIdx !== -1 && endIdx !== -1) {
            const content = input.substring(startIdx + condStart.length, endIdx);
            if (condition) {
                input = input.substring(0, startIdx) + content + input.substring(endIdx + condEnd.length);
            } else {
                input = input.substring(0, startIdx) + input.substring(endIdx + condEnd.length);
            }
            startIdx = input.toLowerCase().indexOf(condStart.toLowerCase());
            endIdx = input.toLowerCase().indexOf(condEnd.toLowerCase());
        }
        
        // Also handle without space: {{/Selected}}
        const condEndNoSpace = `{{/${key}}}`;
        startIdx = input.toLowerCase().indexOf(condStart.toLowerCase());
        endIdx = input.toLowerCase().indexOf(condEndNoSpace.toLowerCase());
        
        while (startIdx !== -1 && endIdx !== -1) {
            const content = input.substring(startIdx + condStart.length, endIdx);
            if (condition) {
                input = input.substring(0, startIdx) + content + input.substring(endIdx + condEndNoSpace.length);
            } else {
                input = input.substring(0, startIdx) + input.substring(endIdx + condEndNoSpace.length);
            }
            startIdx = input.toLowerCase().indexOf(condStart.toLowerCase());
            endIdx = input.toLowerCase().indexOf(condEndNoSpace.toLowerCase());
        }
        
        return input;
    }

    /**
     * Helper method to replace all case-insensitive occurrences
     * @param {string} input Input string
     * @param {string} search Search string
     * @param {string} replacement Replacement string
     * @returns {string} String with all occurrences replaced
     */
    replaceAllCaseInsensitive(input, search, replacement) {
        let idx = 0;
        while (true) {
            const found = input.toLowerCase().indexOf(search.toLowerCase(), idx);
            if (found === -1) break;
            input = input.substring(0, found) + replacement + input.substring(found + search.length);
            idx = found + replacement.length;
        }
        return input;
    }

    /**
     * Applies AppView fallback logic to template replacement text using the centralized getTemplate method
     * @param {string} originalText Original placeholder text
     * @param {string} replacementText Current replacement text
     * @param {Object.<string, PreprocessedTemplate>} preprocessedTemplates Dictionary of preprocessed templates
     * @param {string|null} appView The application view name
     * @returns {string} Updated replacement text with AppView logic applied
     */
    applyAppViewLogicToReplacement(originalText, replacementText, preprocessedTemplates, appView) {
        // Check if the original text is a placeholder that should use AppView fallback logic
        // Extract placeholder name from {{PlaceholderName}} format
        const placeholderName = this.extractPlaceholderName(originalText);
        if (!placeholderName) {
            return replacementText;
        }

        // Use the centralized getTemplate method for consistent AppView logic
        // First get the appSite from the template key pattern
        const sampleKey = Object.keys(preprocessedTemplates)[0];
        if (!sampleKey) {
            return replacementText;
        }
            
        const parts = sampleKey.split('_');
        if (parts.length < 2) {
            return replacementText;
        }
            
        const appSite = parts[0]; // Extract appSite from the key pattern
        
        const template = this.getTemplate(appSite, placeholderName, preprocessedTemplates, appView, this.appViewPrefix, true);
        
        return template?.originalContent || replacementText;
    }

    /**
     * Extracts placeholder name from {{PlaceholderName}} format
     * @param {string} originalText Original placeholder text
     * @returns {string} Extracted placeholder name or empty string
     */
    extractPlaceholderName(originalText) {
        if (!originalText || !originalText.startsWith('{{') || !originalText.endsWith('}}')) {
            return '';
        }
        
        return originalText.substring(2, originalText.length - 2).trim();
    }
}

File: rust/Assembler/src/engine/engine_preprocess.rs
use crate::model::model_preprocess::{PreprocessedTemplate, ReplacementType};
use crate::common::common_util::CommonUtil;
use crate::common::logger::Logger;
use std::collections::HashMap;

/// <summary>
/// PreProcess template engine implementation that only does merging using preprocessed data structures
/// All parsing is done by TemplateLoader, this engine only handles merging
/// </summary>
pub struct EnginePreProcess {
    app_view_prefix: String,
}

impl EnginePreProcess {
    pub fn new(prefix: String) -> Self {
        Self { app_view_prefix: prefix }
    }

    pub fn app_view_prefix(&self) -> &str {
        &self.app_view_prefix
    }
    
    pub fn set_app_view_prefix(&mut self, prefix: String) {
        self.app_view_prefix = prefix;
    }
    
    /// <summary>
    /// Merges templates using preprocessed data structures
    /// This method only does merging using preprocessed data structures - no loading or parsing
    /// </summary>
    /// <param name="app_site">The application site name for template key generation</param>
    /// <param name="app_file">The application file name</param>
    /// <param name="app_view">The application view name (optional)</param>
    /// <param name="preprocessed_templates">Dictionary of preprocessed templates for this specific appSite</param>
    /// <param name="enable_json_processing">Whether to enable JSON data processing</param>
    /// <returns>HTML with placeholders replaced using preprocessed structures</returns>
    pub fn merge_templates(
        &self,
        app_site: &str,
        app_file: &str,
        app_view: Option<&str>,
        preprocessed_templates: &HashMap<String, PreprocessedTemplate>,
        enable_json_processing: bool,
    ) -> String {
        let app_view_str = app_view.unwrap_or("null");
        Logger::debug(&format!("MergeTemplates called: appSite={}, appFile={}, appView={}, enableJson={}",
            app_site, app_file, app_view_str, enable_json_processing), Some("EnginePreProcess"));

        if preprocessed_templates.is_empty() {
            Logger::warn("No preprocessed templates available", Some("EnginePreProcess"));
            return String::new();
        }

        Logger::debug(&format!("Using {} preprocessed templates", preprocessed_templates.len()), Some("EnginePreProcess"));

        // Use the get_template method to retrieve the main template
        let main_preprocessed = self.get_template(
            app_site,
            app_file,
            preprocessed_templates,
            app_view,
            Some(&self.app_view_prefix),
            true,
        );

        if main_preprocessed.is_none() {
            Logger::warn(&format!("Main template not found for appSite={}, appFile={}", app_site, app_file), Some("EnginePreProcess"));
            return String::new();
        }

        let main_preprocessed = main_preprocessed.unwrap();

        Logger::debug(&format!("Main template found, original size: {}", main_preprocessed.original_content.len()), Some("EnginePreProcess"));

        // Start with original content
        let content_html = main_preprocessed.original_content.clone();

        // Apply ALL replacement mappings from ALL templates (TemplateLoader did all the processing)
        let result = self.apply_template_replacements(&content_html, preprocessed_templates, enable_json_processing, app_view, main_preprocessed);

        Logger::debug(&format!("MergeTemplates complete: output size={}", result.len()), Some("EnginePreProcess"));
        result
    }

    /// <summary>
    /// Retrieves a template from the preprocessed templates dictionary based on various scenarios including AppView fallback logic
    /// </summary>
    /// <param name="app_site">The application site name</param>
    /// <param name="template_name">The template name (can be appFile or placeholderName)</param>
    /// <param name="preprocessed_templates">Dictionary of preprocessed templates</param>
    /// <param name="app_view">The application view name (optional)</param>
    /// <param name="app_view_prefix">The application view prefix (optional, uses instance property if not provided)</param>
    /// <param name="use_app_view_fallback">Whether to apply AppView fallback logic</param>
    /// <returns>The template's original content if found, None otherwise</returns>
    fn get_template< 'a>(
        &self,
        app_site: &str,
        template_name: &str,
        preprocessed_templates: &'a HashMap<String, PreprocessedTemplate>,
        app_view: Option<&str>,
        app_view_prefix: Option<&str>,
        use_app_view_fallback: bool,
    ) -> Option<&'a PreprocessedTemplate> {
        if preprocessed_templates.is_empty() {
            return None;
        }

        let view_prefix = app_view_prefix.unwrap_or(&self.app_view_prefix);
        
        // FIRST: Check for AppView-specific template resolution when AppView context is provided
        if use_app_view_fallback {
            if let (Some(app_view), Some(view_prefix)) = (app_view, Some(view_prefix)) {
                if !view_prefix.is_empty() && template_name.to_lowercase().contains(&view_prefix.to_lowercase()) {
                    let app_key = CommonUtil::replace_case_insensitive(template_name, view_prefix, app_view);
                    let fallback_template_key = format!("{}_{}", app_site.to_lowercase(), app_key.to_lowercase());
                    if let Some(fallback_template) = preprocessed_templates.get(&fallback_template_key) {
                        return Some(fallback_template);
                    }
                }
            }
        }

        // SECOND: If no AppView-specific template found, try primary template
        let primary_template_key = format!("{}_{}", app_site.to_lowercase(), template_name.to_lowercase());
        if let Some(primary_template) = preprocessed_templates.get(&primary_template_key) {
            return Some(primary_template);
        }

        None
    }

    /// <summary>
    /// Applies all replacement mappings from all templates - NO processing logic, only direct replacements
    /// </summary>
    fn apply_template_replacements(
        &self,
        content: &str,
        preprocessed_templates: &HashMap<String, PreprocessedTemplate>,
        enable_json_processing: bool,
        app_view: Option<&str>,
        main_template: &PreprocessedTemplate,
    ) -> String {
        Logger::debug(&format!("Starting ApplyTemplateReplacements, initial size: {}", content.len()), Some("EnginePreProcess"));

        let mut result = content.to_string();

        let mut previous;
        let max_passes = 10;
        let mut current_pass = 0;

        loop {
            previous = result.clone();
            current_pass += 1;

            Logger::debug(&format!("Replacement pass {}, current size: {}", current_pass, result.len()), Some("EnginePreProcess"));

            let mut json_placeholder_count = 0;
            let mut slotted_count = 0;
            let mut simple_count = 0;

            // FIRST: Apply JSON placeholder mappings ONLY from the main template (to avoid overwriting component content)
            if current_pass == 1 && enable_json_processing {
                for mapping in main_template.replacement_mappings.iter().filter(|m| matches!(m.r#type, ReplacementType::JsonPlaceholder)) {
                    if result.contains(&mapping.original_text) {
                        Logger::debug(&format!("Applying main template JSON placeholder: {} -> {}", mapping.original_text, mapping.replacement_text), Some("EnginePreProcess"));
                        result = result.replace(&mapping.original_text, &mapping.replacement_text);
                        json_placeholder_count += 1;
                    }
                }
            }

            for template in preprocessed_templates.values() {
                for mapping in template.replacement_mappings.iter().filter(|m| matches!(m.r#type, ReplacementType::SlottedTemplate)) {
                    if result.contains(&mapping.original_text) {
                        let preview_len = std::cmp::min(50, mapping.original_text.len());
                        let preview = &mapping.original_text[..preview_len];
                        Logger::debug(&format!("Applying slotted template: {}... -> {} chars", preview, mapping.replacement_text.len()), Some("EnginePreProcess"));
                        result = result.replace(&mapping.original_text, &mapping.replacement_text);
                        slotted_count += 1;
                    }
                }

                for mapping in template.replacement_mappings.iter().filter(|m| matches!(m.r#type, ReplacementType::SimpleTemplate)) {
                    if result.contains(&mapping.original_text) {
                        let replacement_text = self.apply_app_view_logic_to_replacement(&mapping.original_text, &mapping.replacement_text, preprocessed_templates, app_view);
                        let preview_len = std::cmp::min(200, replacement_text.len());
                        let preview = &replacement_text[..preview_len];
                        Logger::debug(&format!("Applying simple template: {} -> replacement text (first 200 chars): {}", mapping.original_text, preview), Some("EnginePreProcess"));
                        result = result.replace(&mapping.original_text, &replacement_text);
                        simple_count += 1;
                    }
                }
            }

            Logger::debug(&format!("Pass {} applied: {} main JSON placeholders, {} slotted, {} simple", current_pass, json_placeholder_count, slotted_count, simple_count), Some("EnginePreProcess"));

            if result == previous || current_pass >= max_passes {
                break;
            }
        }

        Logger::debug(&format!("Replacement complete after {} passes, final size: {}", current_pass, result.len()), Some("EnginePreProcess"));
        result
    }

    /// <summary>
    /// Applies AppView fallback logic to template replacement text using the centralized get_template method
    /// </summary>
    fn apply_app_view_logic_to_replacement(
        &self,
        original_text: &str,
        replacement_text: &str,
        preprocessed_templates: &HashMap<String, PreprocessedTemplate>,
        app_view: Option<&str>,
    ) -> String {
        // If no appView context, use the default replacement text (which already has JSON values baked in)
        if app_view.is_none() || app_view == Some("") {
            return replacement_text.to_string();
        }

        // Extract placeholder name from {{PlaceholderName}} format
        let placeholder_name = Self::extract_placeholder_name(original_text);
        if placeholder_name.is_empty() {
            return replacement_text.to_string();
        }

        // First get the appSite from the template key pattern
        let sample_key = match preprocessed_templates.keys().next() {
            Some(key) => key,
            None => return replacement_text.to_string(),
        };

        let parts: Vec<&str> = sample_key.split('_').collect();
        if parts.len() < 2 {
            return replacement_text.to_string();
        }

        let app_site = parts[0]; // Extract appSite from the key pattern

        // Use GetTemplate to find the AppView-specific template variant
        let app_view_template = self.get_template(app_site, &placeholder_name, preprocessed_templates, app_view, Some(&self.app_view_prefix), true);

        // If no AppView-specific template found, use the default replacement text
        if app_view_template.is_none() {
            return replacement_text.to_string();
        }

        // Return the AppView-specific template's original content (which already has JSON baked in by LoaderPreProcess)
        app_view_template.unwrap().original_content.clone()
    }

    /// <summary>
    /// Extracts placeholder name from {{PlaceholderName}} format
    /// </summary>
    fn extract_placeholder_name(original_text: &str) -> String {
        if !original_text.starts_with("{{") || !original_text.ends_with("}}") {
            return String::new();
        }
        
        original_text[2..original_text.len() - 2].trim().to_string()
    }
}
